Elasticsearch: Apply ad-hoc filters to annotation queries

mikelv92 opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

Currently the annotation query is not filtered by Ad-Hoc filters in Elasticsearch. This PR adds the Ad-Hoc filters to the annotation query.

**Why do we need this feature?**

Annotation queries are not affected by ad-hoc filters: Grafana still shows annotations that do not match the filter.

**Who is this feature for?**

This feature is for users that are using Annotations from an Elasticsearch datasource and want to filter them.

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #9599 

**Special notes for your reviewer:**

Steps to test it:
- Create an annotation using the `gdev-elasticsearch` datasource
- Create an ad-hoc variable using the `gdev-elasticsearch` datasource
- In the dashboard, set a filter in the ad-hoc variable that you created 
  - The annotations should be filtered too

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

mikelv92 committed:
Elasticsearch: filter annotations

mikelv92 committed:
fix broken tests

mikelv92 committed:
add a new test case

ivanahuckova commented:
@mikelv92 Thank you very much for your contribution. Really appreciated. I will review the PR.

@grafana/observability-logs  - for the context:
I was checking with the dashboard team to ensure that applying ad hoc filters to annotation queries is the correct behavior. They mentioned a couple of things:

1. At one point, we should migrate annotation queries from the deprecated implementation to a new one that uses `this.query`.
2. It is up to the data source to decide what to do with ad hoc filters.

Considering that :
1. This issue has quite a lot of upvotes and was planned to work on in the past
2. After migration, annotation queries would be run through `this.query`, which applies ad hoc filters
3. Data sources have freedom in this

I think that this change makes sense and I will review this PR. But we should really migrate this sooner rather than later and fix typings.

ivanahuckova commented on public/app/plugins/datasource/elasticsearch/datasource.test.ts:
```diff
@@ -1695,6 +1707,61 @@ describe('ElasticDatasource using backend', () => {
         );
       });
+
+      it('should process annotation request using dashboard adhoc variables', async () => {
```
🚀

ivanahuckova commented on public/app/plugins/datasource/elasticsearch/datasource.ts:
```diff
@@ -260,10 +262,15 @@ export class ElasticDatasource
     const timeEndField = annotation.timeEndField || null;
+    const dashboard = options.dashboard;
+
+    const adhocVariables = dashboard.getVariables().filter(v => v.type === "adhoc") as AdHocVariableModel[];
```
```suggestion
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    const adhocVariables = dashboard.getVariables().filter(v => v.type === "adhoc") as AdHocVariableModel[];
```

ivanahuckova replied:
We have a lint rule to not be able to use assertions. In this case, it is needed and correct way to asset `AdHocVariableModel[]` so you'll need to add `// eslint-disable-next-line @typescript-eslint/consistent-type-assertions`

ivanahuckova commented on public/app/plugins/datasource/elasticsearch/datasource.ts:
```diff
@@ -296,6 +303,12 @@ export class ElasticDatasource
 
     const queryInterpolated = this.interpolateLuceneQuery(queryString);
+
+    let finalQuery = queryInterpolated;
+    filters.forEach(filter => {
+      finalQuery = addAddHocFilter(finalQuery, filter);
+    })
```
```suggestion
    const finalQuery = this.addAdHocFilters(queryInterpolated, filters);
```

I think this can be simplified to the code above - to use  `this.addAdHocFilters` that does the same thing.

ivanahuckova commented on public/app/plugins/datasource/elasticsearch/types.ts:
```diff
@@ -1,4 +1,5 @@
 import { DataSourceJsonData } from '@grafana/data';
+import {DataSourceRef} from '@grafana/schema'
```
You'll probably need to run `yarn prettier:write` to fix the linting.

ivanahuckova reviewed (commented):
Really great job, left some comments bellow.

mikelv92 committed:
allow type assertion

There is a lint rule to not be able to use assertions. In this case, it is needed and correct way to asset AdHocVariableModel[]

Co-authored-by: Ivana Huckova <30407135+ivanahuckova@users.noreply.github.com>

mikelv92 committed:
simplify the code to add adhoc filters

Co-authored-by: Ivana Huckova <30407135+ivanahuckova@users.noreply.github.com>

mikelv92 committed:
lint

mikelv92 commented:
Thanks for the review @ivanahuckova . I made the changes according to your recommendations 🙂

ivanahuckova reviewed (approved):
LGTM, thank you so much for another great contribution! Really appreciated. 🧡
